codice/ddf
# DDF-2180: Provide FTPS support for the FTP Endpoint

#### What does this PR do?
Implements FTPS support for the FTP Endpoint
-Provides support for Explicit FTPS
-Provides configuration for 1 and 2 way SSL authentication
-Wait for connections to close before updating server with new configuration 
-Downgrade Apache Mina to 2.0.6. This is because Apache FTPServer has not been updated to newer versions of Apache Mina. This fixes OutOfMemoryError being caught incorrectly, and also gets rid of some erroneous stack traces.
#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged; if a component team is listed, at least one of its members needs to approve)?
@troymohl @mweser @emmberk @lcrosenbu @ricklarsen @stustison 
#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@jaymcnallie
@kcwire
#### How should this be tested?
-Build and run an instance of DDF
-Install `catalog-core-ftp` feature through the Admin UI under DDF Catalog

Insecure connection
-From the command line, `ftp localhost 8021`

1 way authentication
-Ensure `Client Authentication` is set to `Want` in the FTP Endpoint configuration.
-In FileZilla (or another FTPS client of your choice) setup a new site configuration (see screenshots)
-Input password `admin` and accept the Certificate in the pop-up window
-Verify in the FileZilla message log window that TLS was initialized.
-You should be able to connect insecurely using FTP as well, by default.

2 way authentication
-In FTP Endpoint configuration, change `Client Authentication` from Want to Need.
-Download CyberDuck. CyberDuck FTPS client supports client authentication (FileZilla doesn't), or you can use `curl`. `curl` command is `curl --insecure -v -S --ftp-ssl -ftp-method singlecwd -E /path/to/client/cert:certPassword ftp://admin:admin@localhost:8021`. You _should_ be able to use the auto generated `localhost.p12` cert generated after initial startup in `etc/certs/`
-Accept server certificate in popup
-Select client certificate (needs to be trusted by server, `localhost.p12` should already be trusted) to send to server in next popup. The client certificate should be in your systems keychain manager in its Certificates category.

#### Any background context you want to provide?
FTPS is required for encrypting communications between server and client using SSL/TLS.
#### What are the relevant tickets?
DDF-2180
#### Screenshots (if appropriate)
In FileZilla under File->Site Manager
![screen shot 2016-06-14 at 3 25 19 pm](https://cloud.githubusercontent.com/assets/17572782/16061860/5b0d0f2a-3244-11e6-8e09-f7f3b6c36f40.png)

In CyberDuck
![screen shot 2016-06-16 at 2 00 38 pm](https://cloud.githubusercontent.com/assets/17572782/16133120/1aa2ed28-33cb-11e6-997d-57e6169c51d0.png)


#### Checklist:
